src/lib/tuning/tuner.cpp

```diff
@@ -0,0 +1,199 @@
+  _time_budget_exceeded = false;
+
+  _evaluate_task->schedule();
+  _select_task->schedule();
+  _execute_task->schedule();
+}
+
+bool Tuner::is_running() const {
+  return (_evaluate_task && !_evaluate_task->is_done()) || (_select_task && !_select_task->is_done()) ||
+         (_execute_task && !_execute_task->is_done());
+}
+
+Tuner::Status Tuner::status() const { return _status; }
+
+void Tuner::wait_for_completion() {
+  if (_evaluate_task) {
+    _evaluate_task->join();
+  }
+  if (_select_task) {
+    _select_task->join();
+  }
+  if (_execute_task) {
+    _execute_task->join();
+  }
+}
+
+void Tuner::_evaluate() {
+  LOG_INFO("Begin tuning evaluation phase...");
+
+  const auto begin = RuntimeClock::now();
+
+  _choices.clear();
+  for (const auto& evaluator : _evaluators) {
+    evaluator->evaluate(_choices);
+
+    const auto runtime = RuntimeClock::now() - begin;
+    if (runtime > _remaining_time_budget) {
+      LOG_INFO("Interrupt evaluation phase: Time budget exceeded.");
+      _status = Status::Timeout;
+      return;
+    }
+    if (runtime > _evaluate_time_budget) {
+      LOG_INFO("Interrupt evaluation phase: Evaluate time budget exceeded.");
+      _status = Status::EvaluationTimeout;
+      return;
+    }
+  }
+  _remaining_time_budget -= RuntimeClock::now() - begin;
+
+  _log_choices();
+  LOG_INFO("Tuning evaluation phase completed.");
+}
+
+void Tuner::_select() {
+  if (_status != Status::Running) {
+    LOG_INFO("Skip tuning selection phase because previous step exceeded the time budget.");
+    return;
+  }
+  LOG_INFO("Begin tuning selection phase...");
+
+  const auto begin = RuntimeClock::now();
+
+  _operations = _selector->select(_choices, _cost_budget);
+
+  const auto runtime = RuntimeClock::now() - begin;
+  if (runtime > _remaining_time_budget) {
+    LOG_INFO("Interrupt selection phase: Time budget exceeded.");
+    _status = Status::Timeout;
+    return;
+  }
+  if (runtime > _select_time_budget) {
+    LOG_INFO("Interrupt selection phase: Select time budget exceeded.");
+    _status = Status::SelectionTimeout;
+    return;
+  }
+  _remaining_time_budget -= runtime;
+
+  _log_operations();
+  LOG_INFO("Tuning selection phase completed.");
+}
+
+void Tuner::_execute() {
+  if (_status != Status::Running) {
+    LOG_INFO("Skip tuning execution phase because previous step exceeded the time budget.");
+    return;
+  }
+  LOG_INFO("Begin tuning execution phase...");
+
+  const auto begin = RuntimeClock::now();
+
+  for (auto& operation : _operations) {
+    operation->execute();
+
+    const auto runtime = RuntimeClock::now() - begin;
+    if (runtime > _remaining_time_budget) {
+      LOG_INFO("Interrupt execution phase: Time budget exceeded.");
+      _status = Status::Timeout;
+      return;
+    }
+    if (runtime > _execute_time_budget) {
+      LOG_INFO("Interrupt execution phase: Execute time budget exceeded.");
+      _status = Status::ExecutionTimeout;
+      return;
+    }
```

Comment:
You have a lot of code duplication for checking the time budget. Also, I haven‘t seen you using the `_status` variable except for checking if it is `Running`. If we had a single task, you could simply return early.

Reply:
see above.